### PR TITLE
[receiver/dockerstats] support pids_stats metrics

### PR DESCRIPTION
**Pull request by sigilioso**
**Description:** <Describe what has changed.>

Adds support for `pids_stats` metrics.

pids_stats metrics were introduced in [docker engine API v1.23](https://github.com/moby/moby/blob/1261fe69a3586bb102182aa885197822419c768c/docs/api/version-history.md?plain=1#L612) and that's the reason why docker's API default version has been bumped to 1.23. However, the minimum supported version has not been updated because the metrics are optional and reported only when pids cgroup is supported and the kernel is >= 4.3. On top of that, `pids_stats` may be reported even if the API request is performed using previous versions. Example:

```
curl --unix-socket /var/run/docker.sock -H "Content-Type: application/json" http://1.22/containers/78de07328afff50a9777b07dd36a28c709dffe081baaf67235db618843399643/stats?stream=false |jq
{
(...)
  "pids_stats": {
    "current": 1,
    "limit": 2192
  },
(...)
}
```

**Link to tracking Issue:** #21041

**Testing:**

`ScarapeV2` test has been updated:
- Expectation files for the previous tests now include `pids.count` metric (already being there in the corresponding stats json files)
- Included new `ScrapeV2` test case where `pids_stast`metrics are not reported.
- Included new `ScrapeV2` tests case where both new metrics `container.pids.count` and `container.pids.max` are reported.

Tests where executed as follows:

```
❯ make test
go test -race -timeout 300s -parallel 4 --tags="" ./...
ok      github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dockerstatsreceiver  0.805s
ok      github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dockerstatsreceiver/internal/metadata        0.500s

❯ go test --tags integration ./...
ok      github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dockerstatsreceiver  33.664s
ok      github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dockerstatsreceiver/internal/metadata        0.404s
```

**Documentation:**
New metrics documentation was generated using `mdatagen`

### CODE REVIEW

File: receiver/dockerstatsreceiver/metadata.yaml [hunk 1]
```diff
@@ -628,3 +628,19 @@ metrics:
...
     attributes:
       - interface
+
+  # Pids
+  container.pids.count:
+    enabled: false
+    description: "Number of pids in the container's cgroup."
+    extended_documentation: "It requires docker API 1.23 or higher and kernel version >= 4.3 with pids cgroup supporeted. [More docs](https://www.kernel.org/doc/Documentation/cgroup-v1/pids.txt)"
+    unit: "{pids}"
+    gauge:
```
Comment (jamesmoessis) [inline receiver/dockerstatsreceiver/metadata.yaml:638]: I believe these new metrics should be an up down counter (non-monotonic sum) instead of a gauge. Similar to many of the memory metrics elsewhere in this component. 

This is according to the semantic convention spec for similar metric `system.processes.count`: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/system-metrics.md#systemprocesses---aggregate-system-process-metrics

File: receiver/dockerstatsreceiver/metadata.yaml [hunk 2]
```diff
@@ -628,3 +628,19 @@ metrics:
       aggregation: cumulative
     attributes:
       - interface
+
+  # Pids
+  container.pids.count:
```
Comment (jamesmoessis) [inline receiver/dockerstatsreceiver/metadata.yaml:633]: Related to my comment about the semantic convention: also consider the naming of the metric and try and somewhat align it to the semantic convention. Do you think `container.processes.count` is better for alignment? I'll let you make that call, but it's something to consider.

File: receiver/dockerstatsreceiver/receiver.go [hunk 3]
```diff
@@ -126,6 +126,7 @@ func (r *receiver) recordContainerStats(now pcommon.Timestamp, containerStats *d
 	r.recordMemoryMetrics(now, &containerStats.MemoryStats)
 	r.recordBlkioMetrics(now, &containerStats.BlkioStats)
 	r.recordNetworkMetrics(now, &containerStats.Networks)
+	r.recordPidsMetrics(now, &containerStats.PidsStats)
```
[the remark this answers is not in the document]
Comment (sigilioso) [reply receiver/dockerstatsreceiver/receiver.go:129]: Since [PidStats](https://github.com/moby/moby/blob/v23.0.4/api/types/stats.go#L159) is not defined as a pointer in docker types it will be an empty struct, therefore this pointer cannot be nil even if `pids_stats` are not reported.

The stats response in the new test case [pids_no_stats](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21111/commits/7526d761f8709276f48869cd32ab591cf23ff329#diff-26382ae6339e56dd697451b6cc6c37f948cf8712cf30621a74ea1405c531a514R208-R210) does not include `pids_stats` metrics, showing this use-case.

File: receiver/dockerstatsreceiver/receiver_test.go [hunk 4]
```diff
@@ -149,26 +151,44 @@ func TestScrapeV2(t *testing.T) {
...
 
+	noPidsStatsEngineMock, err := dockerMockServer(&map[string]string{
+		"/v1.23/containers/json":                          filepath.Join(mockFolder, "no_pids_stats", "containers.json"),
+		"/v1.23/containers/" + containerIDs[0] + "/json":  filepath.Join(mockFolder, "no_pids_stats", "container.json"),
+		"/v1.23/containers/" + containerIDs[0] + "/stats": filepath.Join(mockFolder, "no_pids_stats", "stats.json"),
+	})
+	assert.NoError(t, err)
+	defer noPidsStatsEngineMock.Close()
+
+	pidsStatsMaxEngineMock, err := dockerMockServer(&map[string]string{
```
[the remark this answers is not in the document]
Comment (sigilioso) [reply receiver/dockerstatsreceiver/receiver_test.go:184]: In this test case both metrics are returned in the [stats response](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21111/commits/7526d761f8709276f48869cd32ab591cf23ff329#diff-94b71c17cc91a1f2a3871d2ba3016d707e1f3e6cc42fd489610dc490b56d9790R4-R6):

```json
  "pids_stats": {
    "current": 1,
    "limit": 2192
  },
```

(`limit` is only reported if a limit is defined, when it is unlimited it is not reported)

Then, the corresponding [expectation file](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21111/commits/7526d761f8709276f48869cd32ab591cf23ff329#diff-139c82de41b2cc8f6dfd00339eeba3e491bc13965150567ae250265cd9ea4ea2R397-R412) also include both metrics.

On the other hand, the existing test cases' stats response returned the `current` value only and that is the only value added in the expectation file.

File: receiver/dockerstatsreceiver/metadata.yaml [hunk 5]
```diff
@@ -628,3 +628,19 @@ metrics:
...
+    enabled: false
+    description: "Number of pids in the container's cgroup."
+    extended_documentation: "It requires docker API 1.23 or higher and kernel version >= 4.3 with pids cgroup supporeted. [More docs](https://www.kernel.org/doc/Documentation/cgroup-v1/pids.txt)"
```
[the remark this answers is not in the document]
Comment (sigilioso) [reply receiver/dockerstatsreceiver/metadata.yaml]: Thanks for pointing out, fixed here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21111/commits/a0c19b5a17e6a763e420cd1e7a96aff384deda21

File: receiver/dockerstatsreceiver/receiver_test.go [hunk 6]
```diff
@@ -149,26 +151,44 @@ func TestScrapeV2(t *testing.T) {
...
+		"/v1.23/containers/json":                          filepath.Join(mockFolder, "pids_stats_max", "containers.json"),
+		"/v1.23/containers/" + containerIDs[3] + "/json":  filepath.Join(mockFolder, "pids_stats_max", "container.json"),
+		"/v1.23/containers/" + containerIDs[3] + "/stats": filepath.Join(mockFolder, "pids_stats_max", "stats.json"),
+	})
+	assert.NoError(t, err)
+	defer pidsStatsMaxEngineMock.Close()
+
```
[the remark this answers is not in the document]
Comment (sigilioso) [reply receiver/dockerstatsreceiver/receiver_test.go]: Thanks for the suggestion! I didn't consider changing the test structure but it makes a lot of sense. I moved the mock definition inside the test case struct as suggested. Since it is in a function and the server is created _inside_ of each `t.Run` I think it is safe closing it inside as well. Please, let me know your thoughts.
Changed here: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21111/commits/74f6e9ac38eabc17a8e11774afcde2211183a05a

File: receiver/dockerstatsreceiver/metadata.yaml (hunk 2 above)
Comment (sigilioso) [reply receiver/dockerstatsreceiver/metadata.yaml:633]: Apparently `processes` is more widely used and is more aligned with the semantic convention, but taking into account that the value is named `pids` in [cgroups](https://www.kernel.org/doc/Documentation/cgroup-v1/pids.txt) and docker API does not change it, maybe keeping `pids` here makes new metrics easier to identify and understand. WDYT?

File: receiver/dockerstatsreceiver/receiver_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/receiver_test.go]: Thanks for improving that, it looks good.

File: receiver/dockerstatsreceiver/metadata.yaml (hunk 2 above)
Comment (jamesmoessis) [reply receiver/dockerstatsreceiver/metadata.yaml:633]: Yeah, I tend to agree. It's reasonable to keep it as `pids` here.

File: receiver/dockerstatsreceiver/metadata.yaml [hunk 7]
```diff
@@ -628,3 +628,24 @@ metrics:
...
+    enabled: false
+    description: "Number of pids in the container's cgroup."
+    extended_documentation: "It requires docker API 1.23 or higher and kernel version >= 4.3 with pids cgroup supported. [More docs](https://www.kernel.org/doc/Documentation/cgroup-v1/pids.txt)"
+    unit: "{pids}"
+    sum:
+      value_type: int
+      aggregation: cumulative
+      monotonic: false
+
+  container.pids.max:
```
Comment (jamesmoessis) [inline receiver/dockerstatsreceiver/metadata.yaml:643]: Is there a reason you've gone with "max" as the terminology here? Seeing as docker and cgroups report it as "limit" it might make sense for the metric name to be `container.pids.limit`. This is also consistent with other metrics in this component such as `container.memory.usage.limit`.

Comment (sigilioso) [reply receiver/dockerstatsreceiver/metadata.yaml:643]: I took it from [cgroups docs](https://www.kernel.org/doc/Documentation/cgroup-v1/pids.txt):

> In order to use the `pids` controller, set the maximum number of tasks in pids.max (this is not available in the root cgroup for obvious reasons). The number of processes currently in the cgroup is given by pids.current.

But, as you stated, it is reported as limit and that name would be consistent with other metrics and the conventions, so I'm changing it. Thanks again! 🙂

Comment (sigilioso) [reply receiver/dockerstatsreceiver/metadata.yaml]: Changed here: [653ad44](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21111/commits/653ad44e3a77e6d1673c5bf19757ac8dc2567008)

**Review by jamesmoessis — APPROVED** [review]
Looks good to me!

Depending on whether https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21166 gets merged before this, you may have some conflicts to resolve.

**Comment by jamesmoessis** [comment]
@codeboten as the codeowner this looks good - just needs approver/maintainer review.

File: receiver/dockerstatsreceiver/receiver.go [hunk 8]
```diff
@@ -26,7 +26,7 @@ import (
 const defaultResourcesLen = 5
 
 const (
-	defaultDockerAPIVersion         = 1.22
+	defaultDockerAPIVersion         = 1.23
```
Comment (atoulme) [inline receiver/dockerstatsreceiver/receiver.go:29]: Is this change related to the addition of those dimensions?

Comment (atoulme) [reply receiver/dockerstatsreceiver/receiver.go:29]: answering my own question - yes, it is.

**Review by atoulme — APPROVED** [review]
